packages/jsii-pacmak/lib/targets/java.ts

```diff
@@ -1513,12 +1522,211 @@ class JavaGenerator extends Generator {
+          this,
+          asListName,
+          `.append("${asListName}")`,
+          {
+            collection: {
+              kind: spec.CollectionKind.Array,
+              elementtype: param.type,
+            },
+          },
+          param.name,
+        );
+      } else {
+        validate.call(
+          this,
+          param.name,
+          `.append("${param.name}")`,
+          param.type,
+          param.name,
+        );
+      }
+    }
+    this.code.closeBlock();
+
+    function validate(
+      this: JavaGenerator,
+      value: string,
+      descr: string,
+      type: spec.TypeReference,
+      parameterName: string,
+    ) {
+      if (spec.isUnionTypeReference(type)) {
+        validateTypeUnion.call(this, value, descr, type, parameterName);
+      } else {
+        // we must cast Collections to access their methods
+        // if they are nested in a union type,
+        // since that union will be rendered as an Object.
+        const collectionType = type as spec.CollectionTypeReference;
+        if (collectionType.collection.kind === spec.CollectionKind.Array) {
+          validateArray.call(
+            this,
+            `((${this.toJavaType(type)})(${value}))`,
+            descr,
+            collectionType.collection.elementtype,
+            parameterName,
+          );
+        } else if (collectionType.collection.kind === spec.CollectionKind.Map) {
+          validateMap.call(
+            this,
+            `((${this.toJavaType(type)})(${value}))`,
+            descr,
+            collectionType.collection.elementtype,
+            parameterName,
+          );
+        } else {
+          throw new Error(
+            `Unhandled collection kind: ${spec.describeTypeReference(type)}`,
+          );
+        }
+      }
+    }
+
+    function validateArray(
+      this: JavaGenerator,
+      value: string,
+      descr: string,
+      elementType: spec.TypeReference,
+      parameterName: string,
+    ) {
+      const varName = `__idx_${createHash('sha256')
+        .update(descr)
+        .digest('hex')
+        .slice(0, 6)}`;
+      this.code.openBlock(
+        `for (int ${varName} = 0; ${varName} < ${value}.size(); ${varName}++)`,
+      );
+      validate.call(
+        this,
+        `${value}.get(${varName})`,
+        `${descr}.append(".get(").append(${varName}).append(")")`,
+        elementType,
+        parameterName,
+      );
+      this.code.closeBlock();
+    }
+
+    function validateMap(
+      this: JavaGenerator,
+      value: string,
+      descr: string,
+      elementType: spec.TypeReference,
+      parameterName: string,
+    ) {
+      const varName = `__item_${createHash('sha256')
+        .update(descr)
+        .digest('hex')
+        .slice(0, 6)}`;
+      const javaElemType = this.toJavaType(elementType);
+      this.code.openBlock(
+        `for (java.util.Map.Entry<String, ${javaElemType}> ${varName}: ${value}.entrySet())`,
+      );
+      validate.call(
```

Comment:
You'll also want to validate that keys are `String` if the `Map` was cast out of a union... In that case, the compiler will not have verified this...

```suggestion
      );
      this.code.openBlock(`if (!(${varName}.getKey() instanceof String))`);
      this.code.line(`throw new IllegalArgumentException("...")`);
      this.code.closeBlock();
      validate.call(
```